File: LeetCode/Easy/0028-find-the-index-of-the-first-occurrence-in-a-string/0028-find-the-index-of-the-first-occurrence-in-a-string.java
class Solution {
    public int strStr(String haystack, String needle) {
        int i = 0;
        while (i < haystack.length()){
            if (haystack.charAt(i) == needle.charAt(0)){
                if (i + needle.length() > haystack.length()) {
                    return -1;
                    }
                boolean isEqual = true;
                boolean isChanged = false;
                int temp = i;
                for (int j = 1; j < needle.length(); ++j){
                    if (!isChanged && haystack.charAt(i+j) == needle.charAt(0)) {
                        temp = i + j - 1;
                        isChanged = true;
                        }
                    if (haystack.charAt(i+j) != needle.charAt(j)){
                        if (!isChanged) temp = i + j;
                        isEqual = false;
                        break;
                    }
                }
                if (isEqual) return i;
                i = temp;
            }
            ++i;
        }
        return -1;
    }
}

File: LeetCode/Easy/0028-find-the-index-of-the-first-occurrence-in-a-string/0028-find-the-index-of-the-first-occurrence-in-a-string.go
func strStr(haystack string, needle string) int {
    i := 0;
    for i < len(haystack){
        if haystack[i] == needle[0] {
            if i + len(needle) > len(haystack) {
                return -1
            }
            isEqual := true
            isChanged := false
            temp := i;
            for j := 1; j < len(needle); j++ {
                if !isChanged && haystack[i+j] == needle[0] {
                    temp = i + j - 1;
                    isChanged = true;
                }
                if haystack[i+j] != needle[j] {
                    if !isChanged {temp = i + j;}
                    isEqual = false;
                    break;
                }
            }
            if isEqual {return i;}
            i = temp;
        }
        i++;
    }
    return -1;
}